### PR TITLE
Upgrade to wasmtime 9.0

### PR DESCRIPTION
**Pull request by lann**

### CODE REVIEW

File: crates/core/src/limits.rs [hunk 1]
```diff
@@ -17,17 +18,27 @@ impl ResourceLimiterAsync for StoreLimitsAsync {
...
         if can_grow {
             self.memory_consumed =
                 (self.memory_consumed as i64 + (desired as i64 - current as i64)) as u64;
         }
-        can_grow
+        Ok(can_grow)
     }
 
-    async fn table_growing(&mut self, _current: u32, desired: u32, _maximum: Option<u32>) -> bool {
-        !matches!(self.max_table_elements, Some(limit) if desired > limit)
```
Comment (lann) [inline crates/core/src/limits.rs:30]: "Who wrote this nonsense?!" <...checks blame...> "I did, of course."

Comment (lann) [reply crates/core/src/limits.rs:30]: Also I screwed up this logic just now. Fixing + adding tests...

File: crates/core/src/preview1.rs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+use wasi_common_preview1::{
+    dir::{OpenResult, ReaddirCursor, ReaddirEntity},
+    file::{Advice, FdFlags, FileType, Filestat, OFlags},
+    Error, ErrorExt, SystemTimeSpec, WasiDir, WasiFile,
+};
+
+pub struct ReadOnlyDir(pub Box<dyn WasiDir>);
+
+#[async_trait::async_trait]
+impl WasiDir for ReadOnlyDir {
```
Comment (lann) [inline crates/core/src/preview1.rs:12]: @dicej

File: crates/core/src/limits.rs [hunk 3]
```diff
@@ -45,3 +60,29 @@ impl StoreLimitsAsync {
...
+
+    #[tokio::test]
+    async fn test_store_limits_memory() {
+        let mut limits = StoreLimitsAsync {
+            max_memory_size: Some(65536),
+            ..Default::default()
+        };
+        assert!(limits.memory_growing(0, 65536, None).await.unwrap());
+        assert_eq!(limits.memory_consumed, 65536);
+        assert!(!limits.memory_growing(65536, 131072, None).await.unwrap());
```
Comment (rylev) [inline crates/core/src/limits.rs:76]:
```suggestion
        assert!(!limits.memory_growing(65536, 131072, None).await.unwrap());
        assert_eq!(limits.memory_consumed, 65536);
```

File: crates/core/src/store.rs [hunk 4]
```diff
@@ -336,29 +288,42 @@ impl StoreBuilder {
...
+                    }
+                    ctx.push_preopened_dir(dir, path)?;
+                }
+                Wasi::Preview2(ctx) => {
+                    let mut dir =
+                        Box::new(wasmtime_wasi_preview2::dir::Dir::from_cap_std(cap_std_dir)) as _;
+                    if !writable {
+                        dir = Box::new(wasi_preview2::dir::ReadOnlyDir(dir));
+                    }
+                    ctx.push_preopened_dir(dir, path)?;
```
Comment (rylev) [inline crates/core/src/store.rs:322]: nit: can this be moved out of the match statement to make it clear that that's what's happening regardless of the which wasi preview.

File: crates/core/src/preview1.rs [hunk 5]
```diff
@@ -0,0 +1,193 @@
+use std::{any::Any, path::PathBuf};
```
Comment (rylev) [inline crates/core/src/preview1.rs:1]: Would it make sense to write a module doc explaining what these implementations are for?

File: crates/core/src/store.rs (hunk 4 above)
Comment (rylev) [reply crates/core/src/store.rs:322]: Ah sorry! I read them as both trait objects of the same type. Whoops!

**Comment by lann** [comment]
This last patch fixes it for me locally. tldr I had deleted all the new preview2 WasiFile methods but I didn't realize there were a bunch of preview1 methods that were removed with preview2, and that `WasiFile` provides default `Error::badf` impls for most of them...

File: crates/core/src/preview1.rs (hunk 2 above)
Comment (lann) [reply crates/core/src/preview1.rs:15]: ...and adding some wrapper methods that were removed in preview2, apparently.

**Review by dicej — APPROVED** [review]

**Pull request merged**